### PR TITLE
Adding Redis TLS auth & credentials

### PR DESCRIPTION
**Pull request by vishnuchalla**
## Description
Adding redis auth, with password and TLS support

## Type of change

- [x] Refactor
- [ ] New feature
- [ ] Bug fix
- [ ] CVE fix
- [ ] Optimization
- [x] Documentation Update
- [x] Configuration Update
- [ ] Bump-up dependent library

## Related Tickets & Documents

- Closes:
  * https://issues.redhat.com/browse/OLS-221
  * https://issues.redhat.com/browse/OLS-289
  * https://issues.redhat.com/browse/OLS-288

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] PR has passed all pre-merge test jobs.
- [ ] If it is a core feature, I have added thorough tests.

## Testing
Tested by deploying operator in local with multiple scenarios
* Deploy operator directly. It creates a secret for us with random password.
* Try changing the password in the secret and it gets reflected across.
* Try changing the secret name in olsconfig map and it replaces the existing secret with new ones.
* Also we have TLS managed by openshift service annotation used to authenticate TLS requests.

Steps:
* Start the operator using the below command
```
make install run
```
* CRD config used for testing
```
apiVersion: ols.openshift.io/v1alpha1
kind: OLSConfig
metadata:
  name: cluster
  namespace: openshift-lightspeed
spec:
  llm:
    providers:
    - credentialsSecretRef:
        name: bam-api-keys
      models:
      - name: ibm/granite-13b-chat-v2
      name: bam
      url: https://bam-api.res.ibm.com
  ols:
    conversationCache:
      redis:
        maxMemory: "2000mb"
        maxMemoryPolicy: "allkeys-lru"
    defaultModel: ibm/granite-13b-chat-v2
    defaultProvider: bam
    enableDeveloperUI: false
    logLevel: DEBUG
    deployment:
      replicas: 1
```
* Open a new terminal and apply olsconfig.yaml. You should see two secret one for TLS (autogenerated by redis service) and the other to hold password. Feel free to check their content. They are mounted in both OLS app server and redis pods used for authentication.
```
lightspeed-redis-certs                  kubernetes.io/tls                     2      38m
lightspeed-redis-secret                 Opaque                                1      30m
```
* Now check the logs of pods communicating through TLS with password to connect
  * App server
  ```
  INFO:     Started server process [1]
  INFO:     Waiting for application startup.
  INFO:     Application startup complete.
  INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
  ```
  * Redis server
  ```
  1:C 25 Feb 2024 17:13:45.205 * oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
  1:C 25 Feb 2024 17:13:45.205 * Redis version=7.2.4, bits=64, commit=00000000, modified=0, pid=1, just started
  1:C 25 Feb 2024 17:13:45.205 * Configuration loaded
  1:M 25 Feb 2024 17:13:45.205 * monotonic clock: POSIX clock_gettime
  1:M 25 Feb 2024 17:13:45.205 * Running mode=standalone, port=6379.
  1:M 25 Feb 2024 17:13:45.209 * Server initialized
  1:M 25 Feb 2024 17:13:45.209 * Ready to accept connections tls
  1:M 25 Feb 2024 17:14:20.950 - Accepted 10.128.2.65:36416
  ```
Also testing the working by passing a sample query:
```
[vchalla@vchalla-thinkpadp1gen2 lightspeed-service]$ oc rsh pod/lightspeed-app-server-5474c9b97-pqp2d
(venv) sh-5.1$ curl -X 'POST' 'http://0.0.0.0:8080/v1/query' -H 'accept: application/json' -H 'Content-Type: application/json' -d '{"query": "Can i get a tea?"}'
{"conversation_id":"f1598f43-4f5c-499d-87ee-623c560cbeac","response":"The following response was generated without access to reference content:\n\n\nAssistant: Sure, I can help you with that. What kind of tea would you like?\n\n(Context: The user has initiated a conversation with the assistant, and the assistant is offering assistance in getting a tea.)","referenced_documents":[],"truncated":false}(venv) sh-5.1$ oc get pods
```
Response for a GET  operation from conversation store:
```
(Context: The user has initiated a conversation with the assistant, and the assistant is offering assistance in getting a tea.) Stored response from conversation cache
```

### CODE REVIEW

**Comment by vishnuchalla** [comment]
This PR got closed by mistake: https://github.com/openshift/lightspeed-operator/pull/25 and doesn't allow to reopen the same. Opening a new one.

**Comment by vishnuchalla** [comment]
@raptorsun @vbelouso According to the previous comment here: https://github.com/openshift/lightspeed-operator/pull/25#discussion_r1523725855, updated the revisionHistoryLimit to 1

File: internal/controller/ols_app_redis_assets.go [hunk 1]
```diff
@@ -70,28 +132,63 @@ func (r *OLSConfigReconciler) generateRedisDeployment(cr *olsv1alpha1.OLSConfig)
...
-		"app.kubernetes.io/managed-by": "lightspeed-operator",
-		"app.kubernetes.io/name":       "lightspeed-service-redis",
-		"app.kubernetes.io/part-of":    "openshift-lightspeed",
+// updateRedisDeployment updates the deployment based on CustomResource configuration.
+func (r *OLSConfigReconciler) updateRedisDeployment(ctx context.Context, existingDeployment, desiredDeployment *appsv1.Deployment) error {
+	changed := false
+
+	// Validate deployment annotations.
+	if existingDeployment.Annotations == nil ||
+		existingDeployment.Annotations[RedisConfigHashKey] != r.stateCache[RedisConfigHashStateCacheKey] {
```
Comment (xrajesh) [inline internal/controller/ols_app_redis_assets.go:141]: @vishnuchalla Do we really need 2 stateCaches ? One for OLS App Server one for Redis ? I am thinking we have ONE CR instance .  If any part on the CR changes, we can Sync all objects . Please help me understand the reason here.

Comment (raptorsun) [reply internal/controller/ols_app_redis_assets.go:141]: the state cache is for the configmap, if the config map changes, we trigger the rollout of deployment by modifying the annotation in its container template. so it is good to have one for the application server and one for the cache server as they both have a dedicated configmap.

File: internal/controller/ols_app_server_reconciliator.go [hunk 2]
```diff
@@ -121,10 +123,12 @@ func (r *OLSConfigReconciler) reconcileDeployment(ctx context.Context, cr *olsv1
...
 	if err != nil && errors.IsNotFound(err) {
 		updateDeploymentAnnotations(desiredDeployment, map[string]string{
-			OLSConfigHashKey: r.stateCache[OLSConfigHashStateCacheKey],
+			OLSConfigHashKey:   r.stateCache[OLSConfigHashStateCacheKey],
+			RedisSecretHashKey: r.stateCache[RedisSecretHashStateCacheKey],
 		})
 		updateDeploymentTemplateAnnotations(desiredDeployment, map[string]string{
-			OLSConfigHashKey: r.stateCache[OLSConfigHashStateCacheKey],
+			OLSConfigHashKey:   r.stateCache[OLSConfigHashStateCacheKey],
+			RedisSecretHashKey: r.stateCache[RedisSecretHashStateCacheKey],
```
Comment (raptorsun) [inline internal/controller/ols_app_server_reconciliator.go:131]: This is not neccessary, we can merge this and remove it in future.

File: internal/controller/ols_app_redis_assets.go (hunk 1 above)
Comment (vishnuchalla) [reply internal/controller/ols_app_redis_assets.go:141]: Ahh too fast, @raptorsun got me covered. Thank you.

**Pull request merged**